### DISCUSSION TITLE
How to free pointers within an on-stack message?

### DISCUSSION THREAD

**Discussion by hu6360567** [protobuf-c/protobuf-c#597] [Q&A] [answered]
In order to simplify memory management on nested messages, I'd like to use `<Message>__free_unpacked(<Message> *message, ProtobufCAllocator *allocator)` to free memory.

But it crashes for messages on the stack.
For now, I'm using a macro from [blog](https://tia.mat.br/posts/2015/05/01/initializing_a_heap_allocated_structure_in_c.html) to allocate C struct on heap with initialization.
```C
void *memdup(const void *src, size_t sz) {
    void *mem = malloc(sz);
    return mem ? memcpy(mem, src, sz) : NULL;
}

#define ALLOC_INIT(type, ...)   \
        (type *)memdup((type[]){ __VA_ARGS__ }, sizeof(type))
```

Is there any other way to allocate message on heap or api to free pointers within an on-stack message?

**Comment by lipnitsk** [comment, marked as answer]
You can't mix stack and heap allocation and rely on the default allocator in __free_unpacked to work properly. Everything needs to be allocated on the heap with one allocator for __free_unpacked to recursively free the message.

Of course, you could design a custom allocator and if it's able to tell how to free each pointer that's passed to it then it should work.
